mftb.md

```diff
@@ -0,0 +1,351 @@
+      - [`Thaw`](#thaw)
+      - [`SetMetadata`](#setmetadata)
+  - [Architecture](#architecture)
+    - [Actions and Reducers Drawbacks](#actions-and-reducers-drawbacks)
+  - [Closing Note](#closing-note)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Purpose of Specification
+
+Typically a specification would describe all the constraints that affect how one integrates with the
+specified software. In the context of Ethereum, [EIPs](https://eips.ethereum.org/) have been
+essential to the creation of a wide range of tools and a web of interconnected contracts. This has
+been possible because Ethereum contract execution happens on-chain. The creators of tools, dependent
+contracts, wallets and other such programs need only be aware of the types and capabilities of their
+target contract. **This is not true of Mina**. To interact with a Mina contract, one needs to run
+the contract code themselves. Each instruction affects how contract results are ultimately proven.
+Interacting with a contract whose method signatures abide by a given spec is not necessarily aligned
+with the desired behavior; in the world of Mina, the implementation is––more or less––the spec. This
+complicates the creation of tools that need to interact with Mina contracts, as these tools need the
+actual contract.
+
+### Long-term Consideration
+
+Some missing pieces may need to come into place before we can design contract specs which solve the
+problems that specs are meant to solve. Some ideas regarding what might be helpful:
+
+- A system for extracting and sharing metadata about the provable properties of a contract.
+- A tool which asserts that a given unknown contract has the specified properties. This would enable
+  the creation and confirmation of "spec"-compliant contracts in the Mina sense of the word "spec."
+
+## Near-term Recommendation
+
+Speccing via TypeScript interfaces alone won't necessarily benefit the community beyond serving as
+inspiration for their own implementations. That being said, there is still an unmet need to provide
+developers a happy path for managing fungible tokens on Mina. To deliver a worthwhile fungible-token
+solution, we could create and deploy a contract and corresponding service that satisfy common use
+cases. Although not a "spec", this contract and service could provide common APIs with which
+developers could manage fungible tokens and administrate those tokens from their own,
+use-case-specific contracts. This solution could serve as a stopgap while we flesh out how the
+protocol itself can eventually satisfy all implementation requirements (the current shortcomings of
+which we touch on below). This could-be stopgap service will be the focus of the remainder of this
+document.
+
+## Design
+
+### Goals
+
+- **Wallet-friendly**: wallets should be able to display and prove data of `FungibleTokenbase`s
+  without needing to dynamically import 3rd party contract code.
```

Comment:
1. Where is the account balance stored?
2. How can I prove that balance is correct without the contract code? I won't be able to generate a valid proof.

Mina seems geared towards a micro-service architecture. I suspect it is necessary to encapsulate the custom code for tokens behind a standardized REST or RPC API that the generic wallet can invoke without needing to know the implementation details. Otherwise, every token will have to be compiled into the code of the wallet which won't scale.

Only other alternative I can think of is to have a registry of addressable contract code that is fetched on the fly and evaluated by the wallet. That is a big security risk though, would need to have a solid sandbox for running untrusted code.

Reply:
I 100% agree. In the case of this proposal, the Fungible Tokenbase contract code would be depended upon by wallets, which would use it to prove balances. However, any admin contracts would not be represented within the wallet. Related work to be taken on: https://github.com/MinaFoundation/Core-Grants/pull/12/files